scvi/external/scar/_module.py

```diff
@@ -0,0 +1,460 @@
+    ):
+        super().__init__(
+            n_input=n_input,
+            n_batch=n_batch,
+            n_labels=n_labels,
+            n_hidden=n_hidden,
+            n_latent=n_latent,
+            n_layers=n_layers,
+            n_continuous_cov=n_continuous_cov,
+            dropout_rate=dropout_rate,
+            dispersion=dispersion,
+            log_variational=log_variational,
+            gene_likelihood=gene_likelihood,
+            latent_distribution=latent_distribution,
+            deeply_inject_covariates=deeply_inject_covariates,
+            use_observed_lib_size=use_observed_lib_size,
+            **vae_kwargs,
+        )
+        self.sparsity = sparsity
+        self.ambient_profile = ambient_profile
+
+        cat_list = [n_batch] + list([] if n_cats_per_cov is None else n_cats_per_cov)
+
+        # decoder goes from n_latent-dimensional space to n_input-d data
+        n_input_decoder = n_latent + n_continuous_cov
+        self.decoder = DecoderSCAR(
+            n_input_decoder,
+            n_input,
+            n_cat_list=cat_list,
+            n_layers=n_layers,
+            n_hidden=n_hidden,
+            inject_covariates=deeply_inject_covariates,
+            scale_activation=scale_activation,
+            sparsity=self.sparsity,
+        )
+
+    @auto_move_data
+    def generative(
+        self,
+        z,
+        library,
+        batch_index,
+        cont_covs=None,
+        cat_covs=None,
+        size_factor=None,
+        y=None,
+        transform_batch=None,
+    ):
+        """Runs the generative model."""
+        # TODO: refactor forward function to not rely on y
+        # Likelihood distribution
+        if cont_covs is None:
+            decoder_input = z
+        elif z.dim() != cont_covs.dim():
+            decoder_input = torch.cat(
+                [z, cont_covs.unsqueeze(0).expand(z.size(0), -1, -1)], dim=-1
+            )
+        else:
+            decoder_input = torch.cat([z, cont_covs], dim=-1)
+
+        if cat_covs is not None:
+            categorical_input = torch.split(cat_covs, 1, dim=1)
+        else:
+            categorical_input = tuple()
+
+        if transform_batch is not None:
+            batch_index = torch.ones_like(batch_index) * transform_batch
+
+        if not self.use_size_factor_key:
+            size_factor = library
+
+        px_scale, px_noise_ratio, px_r, px_rate, px_dropout = self.decoder(
+            self.dispersion,
+            decoder_input,
+            size_factor,
+            batch_index,
+            *categorical_input,
+            y,
+        )
+        if self.dispersion == "gene-label":
+            px_r = F.linear(
+                one_hot(y, self.n_labels), self.px_r
+            )  # px_r gets transposed - last dimension is nb genes
+        elif self.dispersion == "gene-batch":
+            px_r = F.linear(one_hot(batch_index, self.n_batch), self.px_r)
+        elif self.dispersion == "gene":
+            px_r = self.px_r
```

Comment:
Related to earlier comments -- I especially think we don't need this sort of functionality. So I would try to remove.

Reply:
Removed in latest commit